### PR TITLE
feat: use CLUSTER SHARDS 

### LINKED ISSUE #350: Use CLUSTER SHARDS instead of CLUSTER SLOTS.
With the release of Redis 7, the command "CLUSTER SLOTS" will be deprecated, and you should use "CLUSTER SHARDS" instead.

### PR DESCRIPTION
**Pull request by Exca-DK**
closes #350

### CODE REVIEW

File: rueidiscompat/adapter_test.go [hunk 1]
```diff
@@ -129,6 +129,30 @@ func testCluster(resp3 bool) {
 	})
 
 	Describe("Cluster", func() {
+		It("ClusterShards", func() {
+			// TODO Is there other way to ensure that it runs only on server which supports it?? resp3 can be v6.
```
Comment (rueian) [inline rueidiscompat/adapter_test.go:133]: We don't need to worry about that here. Because we use Redis 7 in testing.

File: cluster.go [hunk 2]
```diff
@@ -289,6 +312,74 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
...
+			port = msg["port"].integer
+		}
+		return net.JoinHostPort(ip, strconv.FormatInt(port, 10))
+	}
+
+	groups := make(map[string]group, len(shards.values))
+	for _, v := range shards.values {
+		var master string
+		nodes := v.values[3].values
+		// Not specified that 1st entry is always master
```
Comment (Exca-DK) [inline cluster.go:348]: Only ClusterSlots docs mention that master is always first. Although all example messages in ClusterShards keep master as 1st entry I wasn't sure if it should be assumed.

File: cluster.go [hunk 3]
```diff
@@ -289,6 +312,74 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
 	return groups
 }
 
+// parseShards - map redis shards for each redis nodes/addresses
+// defaultAddr is needed in case the node does not know its own IP
+func parseShards(shards RedisMessage, defaultAddr string) map[string]group {
+	parseNodeEndpoint := func(msg map[string]RedisMessage) string {
+		var ip string
```
Comment (Exca-DK) [inline cluster.go:319]: I kept it the same as in slots although I think that endpoint should be used instead.
> endpoint: The preferred endpoint to reach the node, see below for more information about the possible values of this field.

File: cluster.go (hunk 2 above)
Comment (rueian) [reply cluster.go:348]: Hi @Exca-DK,

We probably need to iterate through the node list and find the first master role.

File: cluster.go [hunk 4]
```diff
@@ -289,6 +312,74 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
...
+		_, isPort := msg["port"]
+		_, isTlsPort := msg["tls-port"]
+		var port int64
+		switch {
+		case isPort && !isTlsPort:
+			port = msg["port"].integer
+		case !isPort && isTlsPort:
+			port = msg["tls-port"].integer
+		default:
+			// TODO which one to choose if both set?
```
Comment (rueian) [inline cluster.go:338]: It should depend on whether the client enables TLS config or not.

Comment (Exca-DK) [reply cluster.go]: Okay I get it, One thing I am unsure of is whether normal ports should even be accepted if TLS has been set.

Comment (rueian) [reply cluster.go]: Yes, I believe it should be accepted if there is no "tls-port" provided.

File: rueidiscompat/command.go [hunk 5]
```diff
@@ -2109,6 +2112,134 @@ func (cmd *ClusterSlotsCmd) Result() ([]ClusterSlot, error) {
...
+		var (
+			start, end int64
+			nodesArr   []ClusterNode
+		)
+		{
+			slots, ok := dict["slots"]
+			if !ok {
+				return &ClusterShardsCmd{err: errors.New("slots not found")}
+			}
+			arr, err := slots.ToArray()
```
Comment (rueian) [inline rueidiscompat/command.go:2135]: We can use `AsIntSlice` instead and ignore the `err`.
```suggestion
			arr, _ := slots.AsIntSlice()
```

File: rueidiscompat/command.go [hunk 6]
```diff
@@ -2109,6 +2112,134 @@ func (cmd *ClusterSlotsCmd) Result() ([]ClusterSlot, error) {
...
+			return &ClusterShardsCmd{err: err}
+		}
+		var (
+			start, end int64
+			nodesArr   []ClusterNode
+		)
+		{
+			slots, ok := dict["slots"]
+			if !ok {
+				return &ClusterShardsCmd{err: errors.New("slots not found")}
```
Comment (rueian) [inline rueidiscompat/command.go:2133]: I think we can just ignore this error.

File: rueidiscompat/command.go [hunk 7]
```diff
@@ -2109,6 +2112,134 @@ func (cmd *ClusterSlotsCmd) Result() ([]ClusterSlot, error) {
...
+			}
+			arr, err := slots.ToArray()
+			if err != nil {
+				return &ClusterShardsCmd{err: err}
+			}
+			if len(arr) != 2 {
```
Comment (rueian) [inline rueidiscompat/command.go:2139]: We can relax the condition to be
```suggestion
			if len(arr) >= 2 {
						start = arr[0]
						end = arr[1]
```

File: rueidiscompat/command.go [hunk 8]
```diff
@@ -2109,6 +2112,134 @@ func (cmd *ClusterSlotsCmd) Result() ([]ClusterSlot, error) {
...
+		val = append(val, ClusterShard{
+			Start: start,
+			End:   end,
+			Nodes: nodesArr,
+		})
+	}
+	return &ClusterShardsCmd{val: val, err: err}
+}
+
+type ClusterShard struct {
```
Comment (rueian) [inline rueidiscompat/command.go:2212]: This struct should be aligned with go-redis

```suggestion
type SlotRange struct {
	Start int64
	End   int64
}

type Node struct {
	ID                string
	Endpoint          string
	IP                string
	Hostname          string
	Port              int64
	TLSPort           int64
	Role              string
	ReplicationOffset int64
	Health            string
}

type ClusterShard struct {
	Slots []SlotRange
	Nodes []Node
}
```

**Comment by rueian** [comment]
image
These two cases should also be covered in tests.

File: rueidiscompat/command.go (hunk 5 above)
Comment (Exca-DK) [reply rueidiscompat/command.go]: The `slots` in `RedisMessage`  are of type `blob string` therefore `AsIntSlice()` doesn't work.

File: rueidiscompat/command.go (hunk 7 above)
Comment (Exca-DK) [reply rueidiscompat/command.go]: goredis seems to keep track of not just first two slots. I aligned it with that now.

File: rueidiscompat/command.go (hunk 5 above)
Comment (rueian) [reply rueidiscompat/command.go]: That's weird. Examples in the document are integers.

File: cluster.go (hunk 3 above)
Comment (rueian) [reply cluster.go:319]: After revisiting the document, I noticed that this logic should be applied to the `endpoint` field, not the `ip` field.

Therefore, we need to change the parse function to be like this:

```go
func parseNodeEndpoint(msg map[string]RedisMessage, defaultAddr string, tls bool) string {
	endpoint := msg["endpoint"].string
	switch endpoint {
	case "":
		return defaultAddr
	case "?":
		return ""
	}

	port := msg["port"].integer
	if tls && msg["tls-port"].integer > 0 {
		port = msg["tls-port"].integer
	}

	return net.JoinHostPort(endpoint, strconv.FormatInt(port, 10))
}
```

File: cluster.go [hunk 9]
```diff
@@ -289,6 +312,76 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
...
+				port = msg["port"].integer
+			}
+		}
+		return net.JoinHostPort(ip, strconv.FormatInt(port, 10))
+	}
+
+	groups := make(map[string]group, len(shards.values))
+	for _, v := range shards.values {
+		var master string
+		nodes := v.values[3].values
```
Comment (rueian) [inline cluster.go:350]: `nodes` may not always be on index 3 anymore.

File: rueidiscompat/command.go (hunk 5 above)
Comment (Exca-DK) [reply rueidiscompat/command.go]: It's based on the provided docker stack.
```
127.0.0.1:7010> CLUSTER SHARDS
1) 1) "slots"
   2) 1) "10923"
      2) "16383"
   3) "nodes"
   ....
```

File: cluster.go [hunk 10]
```diff
@@ -289,6 +312,69 @@ func parseSlots(slots RedisMessage, defaultAddr string) map[string]group {
...
+				dst := parseNodeEndpoint(dict)
+				if dst == "" {
+					continue
+				}
+				g.nodes = append(g.nodes, dst)
+			}
+		}
+		slots := slotsAndNodes["slots"]
+		arr, _ := slots.ToArray()
+		for i := 0; i < len(arr); i += 2 {
```
Comment (rueian) [inline cluster.go:368]: Use `i + 1 < len(arr)` to make sure that the loop won't panic.
```suggestion
		for i := 0; i + 1 < len(arr); i += 2 {
```

File: rueidiscompat/command.go [hunk 11]
```diff
@@ -2109,6 +2110,116 @@ func (cmd *ClusterSlotsCmd) Result() ([]ClusterSlot, error) {
...
+	for _, v := range arr {
+		dict, err := v.ToMap()
+		if err != nil {
+			return &ClusterShardsCmd{err: err}
+		}
+		shard := ClusterShard{}
+		{
+			slots := dict["slots"]
+			arr, _ := slots.ToArray()
+			for i := 0; i < len(arr); i += 2 {
```
Comment (rueian) [inline rueidiscompat/command.go:2128]:
```suggestion
			for i := 0; i + 1 < len(arr); i += 2 {
```

**Comment by rueian** [comment]
Hi @Exca-DK,

It looks good now. Thank you very much for your contribution!

**Pull request merged**